File: travelAgency/Networking/src/main/java/network/protobuffprotocol/ProtoUtils.java
package network.protobuffprotocol;

import app.model.Flight;
import app.model.User;
import app.model.dto.TicketPurchaseDto;

import java.time.LocalDate;
import java.time.LocalTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.List;

public class ProtoUtils {
    public static AppProtobuffs.Request createLoginRequest (String username, String password) {
        AppProtobuffs.User user = AppProtobuffs.User.newBuilder().setUsername(username).setPassword(password).build();
        AppProtobuffs.Request request = AppProtobuffs.Request.newBuilder().setType(AppProtobuffs.Request.Type.LOGIN)
                .setUser(user).build();
        return request;
    }

    public static AppProtobuffs.Request createLogoutRequest(User employee){
        AppProtobuffs.User user = AppProtobuffs.User.newBuilder().setUsername(employee.getUsername()).build();
        AppProtobuffs.Request request = AppProtobuffs.Request.newBuilder().setType(AppProtobuffs.Request.Type.LOGOUT)
                .setUser(user).build();
        return request;
    }

    public static AppProtobuffs.Response createOkResponse(){
        AppProtobuffs.Response response = AppProtobuffs.Response.newBuilder()
                .setType(AppProtobuffs.Response.Type.OK).build();
        return response;
    }

    public static AppProtobuffs.Response createErrorResponse(String text){
        AppProtobuffs.Response response = AppProtobuffs.Response.newBuilder()
                .setType(AppProtobuffs.Response.Type.ERROR)
                .setError(text).build();
        return response;
    }

    public static String getError(AppProtobuffs.Response response){
        String errorMessage = response.getError();
        return errorMessage;
    }

    public static Flight getFlight(AppProtobuffs.Response response) {
        Flight fl = new Flight( response.getFlight().getDestination(), parseStringToLocalDate(response.getFlight().getDate()), LocalTime.parse(response.getFlight().getTime()), response.getFlight().getAirport(), response.getFlight().getNoSeats());
        fl.setId(response.getFlight().getId());
        return fl;
    }

    public static AppProtobuffs.Request createPurchaseRequest(Flight flight, String name, String address, String tourists, String noSeats) {
        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("dd-MM-yyyy");
        String formattedDate = flight.getDate().format(formatter);

        DateTimeFormatter formatterTime = DateTimeFormatter.ofPattern("HH:mm:ss");
        String formattedTime = flight.getHour().format(formatterTime);

        AppProtobuffs.Flight fl = AppProtobuffs.Flight.newBuilder().setId(flight.getId()).setDestination(flight.getDestination()).setDate(formattedDate)
                .setTime(formattedTime).setNoSeats(flight.getNoSeats()).build();
        AppProtobuffs.TicketPurchase tpDto = AppProtobuffs.TicketPurchase.newBuilder().setFlight(fl).setClientName(name).setClientAddress(address)
                .setTouristsNames(tourists).setNoOfTickets(noSeats).build();
        AppProtobuffs.Request request = AppProtobuffs.Request.newBuilder().setType(AppProtobuffs.Request.Type.PURCHASE).setTicket(tpDto).build();
        return request;
    }

    public static AppProtobuffs.Request createFindAllFlightsRequest() {
        AppProtobuffs.Request request = AppProtobuffs.Request.newBuilder().setType(AppProtobuffs.Request.Type.GET_FLIGHTS)
                .build();
        return request;
    }

    public static Iterable<Flight> getAllFlights(AppProtobuffs.Response response) {
        List<Flight> flights = new ArrayList<>();
        for(int i = 0; i < response.getFlightsCount(); i++){
            AppProtobuffs.Flight fl = response.getFlights(i);

            DateTimeFormatter formatter = DateTimeFormatter.ofPattern("dd-MM-yyyy");
            LocalDate localDate = LocalDate.parse(fl.getDate(), formatter);

            DateTimeFormatter formatterTime = DateTimeFormatter.ofPattern("HH:mm:ss");
            LocalTime localTime = LocalTime.parse(fl.getTime(), formatterTime);

            Flight flight = new Flight(fl.getDestination(), localDate, localTime, fl.getAirport(), fl.getNoSeats());
            flight.setId(fl.getId());
            flights.add(flight);
        }
        return flights;
    }

    public static AppProtobuffs.Request createFindFlightsDestDateRequest(String dest, String date) {
        AppProtobuffs.SearchFlights srcFl = AppProtobuffs.SearchFlights.newBuilder().setDestination(dest).setDate(date).build();
        AppProtobuffs.Request request = AppProtobuffs.Request.newBuilder().setType(AppProtobuffs.Request.Type.GET_FLIGHTS_DEST_DATE).setSearchFlights(srcFl)
                .build();
        return request;
    }


    private static LocalDate parseStringToLocalDate(String str){
        String[] parts = str.split("-");
        return LocalDate.of(Integer.parseInt(parts[2]), Integer.parseInt(parts[1]), Integer.parseInt(parts[0]));
    }
}

File: Anul_2/Semestrul_2/Medii de proiectare si programare/mpp-proiect-csharp-andrea-muresan/TravelAgencyProject/NetworkingProto/ProtoUtils.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Model.dto;

namespace NetworkingProto
{
    public class ProtoUtils
    {
        public static Model.dto.UserDto getUserDto(Request request)
        {
            Model.dto.UserDto userDto = new Model.dto.UserDto(request.User.Username, request.User.Password);
            return userDto;
        }
        
        public static Response createOkResponse()
        {
            Response response = new Response { Type = Response.Types.Type.Ok };
            return response;
        }
        
        public static Response createErrorResponse(string message)
        {
            Response response = new Response { Type = Response.Types.Type.Error, Error = message };
            return response;
        }
        
        public static Model.dto.TicketPurchaseDto getTicketPurchaseDto(Request request)
        {
            Model.dto.TicketPurchaseDto tpDto = new Model.dto.TicketPurchaseDto(request.Ticket.Flight.Id, request.Ticket.ClientName, request.Ticket.ClientAddress, request.Ticket.TouristsNames, request.Ticket.NoOfTickets.ToString());
            return tpDto;
        }
        
        public static Response createFindAllFlightsResponse(IEnumerable<Model.Flight> flights)
        {
            Response response = new Response { Type = Response.Types.Type.GetFlights };
            foreach (Model.Flight flight in flights)
            {
                var x = flight.Date;
                String date =  x.ToString("dd-MM-yyyy");;
                String time = x.ToString("HH:mm:ss");
                Flight fl = new Flight
                {
                    Id = flight.Id, Destination = flight.Destination, Date = date, Time = time,
                    Airport = flight.Airport, NoSeats = flight.NoSeats
                };
                
                response.Flights.Add(fl);
            }
            return response;
        }
        
        public static Model.dto.SearchFlightsDto getSearchFlightDto(Request request)
        {
            Model.dto.SearchFlightsDto searchFlightsDto = new Model.dto.SearchFlightsDto(request.SearchFlights.Destination, request.SearchFlights.Date);
            return searchFlightsDto;
        }
        
        public static Response createFindFlightsDestDateResponse(IEnumerable<Model.Flight> flights)
        {
            Response response = new Response { Type = Response.Types.Type.GetFlightsDestDate };
            foreach (Model.Flight flight in flights)
            {
                var x = flight.Date;
                String date =  x.ToString("dd-MM-yyyy");;
                String time = x.ToString("HH:mm:ss");
                Flight fl = new Flight
                {
                    Id = flight.Id, Destination = flight.Destination, Date = date, Time = time,
                    Airport = flight.Airport, NoSeats = flight.NoSeats
                };
                
                response.Flights.Add(fl);
            }
            return response;
        }
        
        
        
        public static Response createFlightResponse(Model.Flight flight)
        {
            var x = flight.Date;
            String date =  x.ToString("dd-MM-yyyy");;
            String time = x.ToString("HH:mm:ss");
            Flight fl = new Flight { Id = flight.Id, Destination = flight.Destination, Date = date, Time = time, Airport = flight.Airport, NoSeats = flight.NoSeats};
            Response response = new Response { Type = Response.Types.Type.Flight , Flight = fl};
            return response;
        }
    }
}